// @loaders.gl/3d-tiles/dist/es5/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TILE3D_REFINEMENT = exports.TILE3D_COLOR_BLEND_MODE = exports.TILE3D_OPTIMIZATION_HINT = exports.TILE3D_CONTENT_STATE = exports.MAGIC_ARRAY = exports.TILE3D_TYPES = exports.TILE3D_TYPE = void 0;
var TILE3D_TYPE = {
  COMPOSITE: 'cmpt',
  POINT_CLOUD: 'pnts',
  BATCHED_3D_MODEL: 'b3dm',
  INSTANCED_3D_MODEL: 'i3dm',
  GEOMETRY: 'geom',
  VECTOR: 'vect'
};
exports.TILE3D_TYPE = TILE3D_TYPE;
var TILE3D_TYPES = Object.keys(TILE3D_TYPE);
exports.TILE3D_TYPES = TILE3D_TYPES;
var MAGIC_ARRAY = {
  BATCHED_MODEL: [98, 51, 100, 109],
  INSTANCED_MODEL: [105, 51, 100, 109],
  POINT_CLOUD: [112, 110, 116, 115],
  COMPOSITE: [99, 109, 112, 116]
};
exports.MAGIC_ARRAY = MAGIC_ARRAY;
var TILE3D_CONTENT_STATE = {
  UNLOADED: 0,
  LOADING: 1,
  PROCESSING: 2,
  READY: 3,
  EXPIRED: 4,
  FAILED: 5
};
exports.TILE3D_CONTENT_STATE = TILE3D_CONTENT_STATE;
var TILE3D_OPTIMIZATION_HINT = {
  NOT_COMPUTED: -1,
  USE_OPTIMIZATION: 1,
  SKIP_OPTIMIZATION: 0
};
exports.TILE3D_OPTIMIZATION_HINT = TILE3D_OPTIMIZATION_HINT;
var TILE3D_COLOR_BLEND_MODE = {
  HIGHLIGHT: 0,
  REPLACE: 1,
  MIX: 2
};
exports.TILE3D_COLOR_BLEND_MODE = TILE3D_COLOR_BLEND_MODE;
var TILE3D_REFINEMENT = {
  ADD: 0,
  REPLACE: 1
};
exports.TILE3D_REFINEMENT = TILE3D_REFINEMENT;
