// @loaders.gl/3d-tiles/dist/es5/tile-3d-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _parse3dTile = require("./parsers/parse-3d-tile");

function parse(_x, _x2, _x3, _x4) {
  return _parse.apply(this, arguments);
}

function _parse() {
  _parse = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee(arrayBuffer, options, context, loader) {
    var tile, byteOffset;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            tile = {};
            byteOffset = 0;
            _context.next = 4;
            return (0, _parse3dTile.parse3DTile)(arrayBuffer, byteOffset, options, context, tile);

          case 4:
            return _context.abrupt("return", tile);

          case 5:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _parse.apply(this, arguments);
}

function parseSync(arrayBuffer, options, context, loader) {
  var tile = {};
  var byteOffset = 0;
  (0, _parse3dTile.parse3DTileSync)(arrayBuffer, byteOffset, options, context, tile);
  return tile;
}

var _default = {
  name: '3D Tiles',
  extensions: ['cmpt', 'pnts', 'b3dm', 'i3dm'],
  mimeType: 'application/octet-stream',
  test: ['cmpt', 'pnts', 'b3dm', 'i3dm'],
  parse: parse,
  parseSync: parseSync,
  binary: true
};
exports["default"] = _default;
