// @deck.gl/geo-layers/dist/es6/tile-layer/utils/tile-cache.js
import Tile from './tile';
import { getTileIndices } from './viewport-util';
export default class TileCache {
  constructor({
    getTileData,
    maxSize,
    maxZoom,
    minZoom,
    onTileLoad,
    onTileError
  }) {
    this._getTileData = getTileData;
    this._maxSize = maxSize;
    this.onTileError = onTileError;
    this.onTileLoad = onTileLoad;
    this._cache = new Map();
    this._tiles = [];

    if (Number.isFinite(maxZoom)) {
      this._maxZoom = Math.floor(maxZoom);
    }

    if (Number.isFinite(minZoom)) {
      this._minZoom = Math.ceil(minZoom);
    }
  }

  get tiles() {
    return this._tiles;
  }

  finalize() {
    this._cache.clear();
  }

  update(viewport) {
    const {
      _cache,
      _getTileData,
      _maxSize,
      _maxZoom,
      _minZoom
    } = this;

    this._markOldTiles();

    const tileIndices = getTileIndices(viewport, _maxZoom, _minZoom);

    if (!tileIndices || tileIndices.length === 0) {
      return;
    }

    _cache.forEach(cachedTile => {
      if (tileIndices.some(tile => cachedTile.isOverlapped(tile))) {
        cachedTile.isVisible = true;
      }
    });

    let changed = false;

    for (let i = 0; i < tileIndices.length; i++) {
      const tileIndex = tileIndices[i];
      const {
        x,
        y,
        z
      } = tileIndex;

      let tile = this._getTile(x, y, z);

      if (!tile) {
        tile = new Tile({
          getTileData: _getTileData,
          x,
          y,
          z,
          onTileLoad: this.onTileLoad,
          onTileError: this.onTileError
        });
        tile.isVisible = true;
        changed = true;
      }

      const tileId = this._getTileId(x, y, z);

      _cache.set(tileId, tile);
    }

    if (changed) {
      const commonZoomRange = 5;

      this._resizeCache(_maxSize || commonZoomRange * tileIndices.length);

      this._tiles = Array.from(this._cache.values()).sort((t1, t2) => t1.z - t2.z);
    }
  }

  _resizeCache(maxSize) {
    const {
      _cache
    } = this;

    if (_cache.size > maxSize) {
      const iterator = _cache[Symbol.iterator]();

      for (const cachedTile of iterator) {
        if (_cache.size <= maxSize) {
          break;
        }

        const tileId = cachedTile[0];
        const tile = cachedTile[1];

        if (!tile.isVisible) {
          _cache.delete(tileId);
        }
      }
    }
  }

  _markOldTiles() {
    this._cache.forEach(cachedTile => {
      cachedTile.isVisible = false;
    });
  }

  _getTile(x, y, z) {
    const tileId = this._getTileId(x, y, z);

    return this._cache.get(tileId);
  }

  _getTileId(x, y, z) {
    return "".concat(z, "-").concat(x, "-").concat(y);
  }

}
