// @math.gl/culling/dist/es6/lib/plane.js
import { Vector3, equals, assert, MathUtils } from 'math.gl';
const scratchPosition = new Vector3();
const scratchNormal = new Vector3();
export default class Plane {
  constructor() {
    let normal = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [0, 0, 1];
    let distance = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
    this.normal = new Vector3();
    this.distance = -0;
    this.fromNormalDistance(normal, distance);
  }

  fromNormalDistance(normal, distance) {
    assert(Number.isFinite(distance));
    this.normal.from(normal).normalize();
    this.distance = distance;
    return this;
  }

  fromPointNormal(point, normal) {
    point = scratchPosition.from(point);
    this.normal.from(normal).normalize();
    const distance = -this.normal.dot(point);
    this.distance = distance;
    return this;
  }

  fromCoefficients(a, b, c, d) {
    this.normal.set(a, b, c);
    assert(this.normal.len() === 1);
    this.distance = d;
    return this;
  }

  clone(plane) {
    return new Plane(this.normal, this.distance);
  }

  equals(right) {
    return equals(this.distance, right.distance) && equals(this.normal, right.normal);
  }

  getPointDistance(point) {
    return this.normal.dot(point) + this.distance;
  }

  transform(matrix4) {
    const normal = scratchNormal.copy(this.normal).transformAsVector(matrix4).normalize();
    const point = this.normal.scale(-this.distance).transform(matrix4);
    return this.fromPointNormal(point, normal);
  }

  projectPointOntoPlane(point) {
    let result = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [0, 0, 0];
    point = scratchPosition.from(point);
    const pointDistance = this.getPointDistance(point);
    const scaledNormal = scratchNormal.copy(this.normal).scale(pointDistance);
    return point.subtract(scaledNormal).to(result);
  }

}
