// @loaders.gl/math/dist/es5/geometry/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GL", {
  enumerable: true,
  get: function get() {
    return _constants.GL;
  }
});
Object.defineProperty(exports, "GL_TYPE", {
  enumerable: true,
  get: function get() {
    return _constants.GL_TYPE;
  }
});
Object.defineProperty(exports, "GLType", {
  enumerable: true,
  get: function get() {
    return _glType["default"];
  }
});
Object.defineProperty(exports, "isGeometry", {
  enumerable: true,
  get: function get() {
    return _isGeometry["default"];
  }
});
Object.defineProperty(exports, "attributeIterator", {
  enumerable: true,
  get: function get() {
    return _attributeIterator["default"];
  }
});
Object.defineProperty(exports, "primitiveIterator", {
  enumerable: true,
  get: function get() {
    return _primitiveIterator["default"];
  }
});
Object.defineProperty(exports, "computeVertexNormals", {
  enumerable: true,
  get: function get() {
    return _computeVertexNormals["default"];
  }
});
Object.defineProperty(exports, "encodeRGB565", {
  enumerable: true,
  get: function get() {
    return _rgb.encodeRGB565;
  }
});
Object.defineProperty(exports, "decodeRGB565", {
  enumerable: true,
  get: function get() {
    return _rgb.decodeRGB565;
  }
});
Object.defineProperty(exports, "concatTypedArrays", {
  enumerable: true,
  get: function get() {
    return _typedArrayUtils.concatTypedArrays;
  }
});
Object.defineProperty(exports, "octEncodeInRange", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeInRange;
  }
});
Object.defineProperty(exports, "octEncode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncode;
  }
});
Object.defineProperty(exports, "octEncodeToVector4", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeToVector4;
  }
});
Object.defineProperty(exports, "octDecodeInRange", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeInRange;
  }
});
Object.defineProperty(exports, "octDecode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecode;
  }
});
Object.defineProperty(exports, "octDecodeFromVector4", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeFromVector4;
  }
});
Object.defineProperty(exports, "octPackFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octPackFloat;
  }
});
Object.defineProperty(exports, "octEncodeFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octEncodeFloat;
  }
});
Object.defineProperty(exports, "octDecodeFloat", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octDecodeFloat;
  }
});
Object.defineProperty(exports, "octPack", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octPack;
  }
});
Object.defineProperty(exports, "octUnpack", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.octUnpack;
  }
});
Object.defineProperty(exports, "compressTextureCoordinates", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.compressTextureCoordinates;
  }
});
Object.defineProperty(exports, "decompressTextureCoordinates", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.decompressTextureCoordinates;
  }
});
Object.defineProperty(exports, "zigZagDeltaDecode", {
  enumerable: true,
  get: function get() {
    return _attributeCompression.zigZagDeltaDecode;
  }
});

var _constants = require("./constants");

var _glType = _interopRequireDefault(require("./gl/gl-type"));

var _isGeometry = _interopRequireDefault(require("./is-geometry"));

var _attributeIterator = _interopRequireDefault(require("./iterators/attribute-iterator"));

var _primitiveIterator = _interopRequireDefault(require("./iterators/primitive-iterator"));

var _computeVertexNormals = _interopRequireDefault(require("./attributes/compute-vertex-normals"));

var _rgb = require("./colors/rgb565");

var _typedArrayUtils = require("./typed-arrays/typed-array-utils");

var _attributeCompression = require("./compression/attribute-compression");
